apps/server/src/modules/files-storage/service/preview.service.ts

```diff
@@ -0,0 +1,110 @@
+import { Injectable, InternalServerErrorException, NotAcceptableException } from '@nestjs/common';
+import { LegacyLogger } from '@src/core/logger';
+import crypto from 'crypto';
+import { subClass } from 'gm';
+import { PassThrough } from 'stream';
+import { S3ClientAdapter } from '../client/s3-client.adapter';
+import { DownloadFileParams, PreviewParams } from '../controller/dto';
+import { FileRecord } from '../entity';
+import { ErrorType } from '../error';
+import { IGetFile, IGetFileResponse } from '../interface';
+import { PreviewOutputMimeTypes } from '../interface/preview-output-mime-types.enum';
+import { FileDtoBuilder, FileResponseBuilder } from '../mapper';
+import { FilesStorageService } from './files-storage.service';
+
+@Injectable()
+export class PreviewService {
+	constructor(
+		private readonly storageClient: S3ClientAdapter,
+		private readonly fileStorageService: FilesStorageService,
+		private logger: LegacyLogger
+	) {
+		this.logger.setContext(PreviewService.name);
+	}
+
+	private checkIfPreviewPossible(fileRecord: FileRecord): void | NotAcceptableException {
+		if (!fileRecord.isPreviewPossible()) {
+			this.logger.warn(`could not generate preview for : ${fileRecord.id} ${fileRecord.mimeType}`);
+			throw new NotAcceptableException(ErrorType.PREVIEW_NOT_POSSIBLE);
+		}
+	}
+
+	public async getPreview(
+		fileRecord: FileRecord,
+		params: DownloadFileParams,
+		previewParams: PreviewParams,
+		bytesRange?: string
+	): Promise<IGetFileResponse> {
+		this.checkIfPreviewPossible(fileRecord);
+
+		const hash = this.createNameHash(params, previewParams);
+		const filePath = ['previews', fileRecord.getSchoolId(), hash].join('/');
+		const fileNameWithoutExtension = fileRecord.name.split('.')[0];
+		const format = this.getFormat(previewParams.outputFormat);
+		const name = `${fileNameWithoutExtension}.${format}`;
+		let file: IGetFile;
+
+		try {
+			file = await this.storageClient.get(filePath, bytesRange);
+		} catch (error) {
+			this.throwIfOtherError(error);
+
+			file = await this.generatePreview(fileRecord, params, previewParams, hash, filePath, bytesRange);
```

Comment:
I think the logic is harder to understand if you do it this way around. I suggest to it this way. Also with some other suggestions for improvement:

```suggestion
if (err instanceof S3ServiceException && err.name === 'NoSuchKey') {
	file = await this.generatePreview(fileRecord, params, previewParams, hash, filePath, bytesRange);
}

throw error;			
```
If you use the instance check with `S3ServiceException` you don't need to disable eslint and the code communicates way clearer what you are doing.

To use `error.name` and `error.Code` should lead to the same result. I read somewhere an example in the AWS docs that suggested to use `name`. Unfortunately I can't find it anymore. If I remember correctly from debugging `error.message` does not work. It is hard to find proper documentation on it. I struggled quite a bit with it in this PR: https://github.com/hpi-schul-cloud/schulcloud-server/pull/4265